changes.go

```diff
@@ -0,0 +1,141 @@
+// Copyright 2015 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package bundlechanges
+
+import (
+	"fmt"
+	"sort"
+
+	"gopkg.in/juju/charm.v6-unstable"
+)
+
+// FromData generates and returns the set of changes required to deploy the
+// given bundle data.
+func FromData(data *charm.BundleData) []*Change {
+	cs := &changeset{}
+	addedServices := handleServices(cs.add, data.Services)
+	addedMachines := handleMachines(cs.add, data.Machines)
+	handleRelations(cs.add, data.Relations, addedServices)
+	handleUnits(cs.add, data.Services, addedServices, addedMachines)
+	return cs.changes
+}
+
+// Change holds a single change required to deploy a bundle.
+type Change struct {
+	// Id is the unique identifier for this change.
+	Id string `json:"id"`
+	// Method is the action to be performed to apply this change.
+	Method string `json:"method"`
+	// Args holds a list of arguments to pass to the method.
+	Args []interface{} `json:"args"`
+	// Requires holds a list of dependencies for this change. Each dependency
+	// is represented by the corresponding change id, and must be applied
+	// before this change is applied.
+	Requires []string `json:"requires"`
+}
+
+// changeset holds the list of changes returned by FromData.
+type changeset struct {
+	changes []*Change
+}
+
+// add is an addChangeFunc that can be used to add a change to this change set.
+func (cs *changeset) add(method string, requires []string, args ...interface{}) *Change {
+	if args == nil {
+		args = make([]interface{}, 0)
+	}
+	if requires == nil {
+		requires = make([]string, 0)
+	}
+	change := &Change{
+		Id:       fmt.Sprintf("%s-%d", method, len(cs.changes)),
+		Method:   method,
+		Args:     args,
+		Requires: requires,
+	}
+	cs.changes = append(cs.changes, change)
+	return change
+}
+
+// addChangeFunc is used to add a change to a change set. The resulting change
+// is also returned.
+type addChangeFunc func(method string, requires []string, args ...interface{}) *Change
```

Comment:
This is defined so that we don't need to repeat the whole signature in the handler functions.
I don't understand "use the add method on the changeset directly". Are you suggesting to pass the changeset to the handlers? I am not sure about that, the handlers only need a function to add a change, not the whole changeset.